drivers/bluetooth/hci: add SiLabs BLE HCI driver

fkokosinski opened the pull request:
This PR adds the SiLabs Bluetooth HCI driver. It also enables this BLE HCI driver on the `efr32bg_sltb010a` board.

This PR depends on the changes introduced in https://github.com/zephyrproject-rtos/hal_silabs/pull/21.

nordicjm reviewed (dismissed):
This needs to follow the rules set out in https://docs.zephyrproject.org/latest/contribute/bin_blobs.html#tainting

fkokosinski commented:
> This needs to follow the rules set out in https://docs.zephyrproject.org/latest/contribute/bin_blobs.html#tainting

Hi @nordicjm, thanks for your comment. Are you referring only to the "New Binary Blobs issue", or is there something else that I'm missing? From what I see, the symbols `CONFIG_TAINT` and `CONFIG_TAINT_BLOBS` are y-selected, and the blobs themselves are downloaded with the `west blobs` command.

nordicjm commented:
Have checked and nothing extra is needed, it is handled by west

fkokosinski commented:
Hi @jhedberg! The binary blobs that this PR depends on [have been approved](https://github.com/zephyrproject-rtos/zephyr/issues/54865#issuecomment-1489788466). Could you take a look at this PR?

fkokosinski commented:
> [... ]and then the HCI driver would be responsible for making those `bt_recv()` calls (like all other HCI drivers).Thoughts?

Agreed. I see that you've already ACK'd these changes, but for the record I addressed your comment by adding the `hci_common_transport_transmit` function implementation, which calls the `bt_recv()` and the `bt_recv_prio()` functions.

rettichschnidi commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -56,6 +56,18 @@ config BT_STM32_IPM
+	select MBEDTLS_ENTROPY_ENABLED
+	select MBEDTLS_ZEPHYR_ENTROPY
+	help
+	  Use Silicon Labs binary Bluetooth library to connect to the
```
It seems like using this component like this:

```
$ west build --board efr32bg_sltb010a zephyr/samples/bluetooth/observer
```

does not print a warning about the build being tainted. I think [it should](https://docs.zephyrproject.org/latest/contribute/bin_blobs.html#tainting).
```
$ grep TAINT build/zephyr/.config
```

fkokosinski replied:
Hi, thanks for commenting. I'm not sure what additional Kconfig changes should I make here, as I couldn't find a preexisting BT driver that does that. I've tested with `esp32c3_devkitm`, which uses blobs as well, and no warning was printed by the build system.

As for taint marking in the Kconfig itself, we're already doing that:
```
% west build -p -b efr32bg_sltb010a samples/bluetooth/observer >
% grep TAINT build/zephyr/.config
CONFIG_TAINT_BLOBS=y
CONFIG_TAINT=y
```

rettichschnidi replied:
> Hi, thanks for commenting. I'm not sure what additional Kconfig changes should I make here, as I couldn't find a preexisting BT driver that does that. I've tested with `esp32c3_devkitm`, which uses blobs as well, and no warning was printed by the build system.

True. Guess fixing this does not have to be part of this PR then.

> As for taint marking in the Kconfig itself, we're already doing that:
> 
> ```
> % west build -p -b efr32bg_sltb010a samples/bluetooth/observer
> % grep TAINT build/zephyr/.config
> CONFIG_TAINT_BLOBS=y
> CONFIG_TAINT=y
> ```
This works when the espressif blobs got fetched, but not when only the silabs ones are.

Try this again after executing the following:
```shell
$ west blobs clean
$ west blobs fetch silabs
```
It will result in a build which does not have TAINT_BLOBS set.

This behaviour seems to boil down to https://github.com/zephyrproject-rtos/zephyr/blob/3d5de8920b971dc3fcb3b2a1504e28737396d25e/scripts/zephyr_module.py#L332 only selecting TAINT_BLOBS when blobs have been found AND `zephyr/Kconfig` exists in the modules. The file can be empty as it is in [the espressif HAL](https://github.com/zephyrproject-rtos/hal_espressif/blob/823e1df5f5b5ae0254f046f6e255cdd7a84a9095/zephyr/Kconfig).

Not sure if changing zephyr_module.py or adding zephyr/Kconfig to the silabs HAL module is the better solution.

fkokosinski replied:
Ok, thanks for elaborating on that. I'll investigate it.

fkokosinski replied:
@rettichschnidi We opted for the latter. Could you retest it and confirm that it works as well on your end?

rettichschnidi replied:
It does. Thanks!

rettichschnidi commented on drivers/bluetooth/hci/slz_hci.c:
```diff
@@ -0,0 +1,142 @@
+	IRQ_CONNECT(AGC_IRQn, 0, AGC_IRQHandler, NULL, 0);
+}
+
+uint32_t hci_common_transport_transmit(uint8_t *data, int16_t len)
```
This function belongs to the OS abstraction layer? If so, please document the input and output ([using Doxygen comments](https://docs.zephyrproject.org/latest/contribute/bin_blobs.html#allowed-dependencies)).

fkokosinski replied:
Thanks for spotting this. I've added the documentation.

rettichschnidi replied:
It looks to me like data[1] has a special meaning (HCI event code? Is there some authoritative documentation which could be linked to?), would mention this specifically.

(I mean, nowadays this is easy to figure out by reading the code, but what when bt_buf_get_evt() gets changed down the road. How can, whoever is tasked to maintaining this code and potentially not part of SiLabs, be sure in many years from now?)

fkokosinski replied:
> It looks to me like data[0] and data[1] have special meanings, would mention those with some extra information (maybe refer to existing documentation?).

I don't think any additional documentation is needed here, as these are covered by the standard. The first byte is the packet type, and in event frames the second byte is for the event code (for LE events). I've changed the variable names to reflect that. Let me know if that's okay.

> (Sorry for being pedantic. Nowadays this is easy to figure out by reading the code, but what when the Zephyr functions get changed down the road. How can, whoever is tasked to maintaining this code and potentially not part of SiLabs, be sure in many years from now how exactly the blobs call this function with?)

No problem :) Thanks for looking at the code and reviewing it.

rettichschnidi replied:
> I don't think any additional documentation is needed here, as these are covered by the standard. The first byte is the packet type, and in event frames the second byte is for the event code (for LE events). I've changed the variable names to reflect that. Let me know if that's okay.

Reasoning seems fine to me. However, I'd describe param data as e.g. "Packet type followed by HCI packet data", as the first byte not actually gets transmitted as the current comment suggests.

fkokosinski replied:
Thanks for your suggestion. I've updated the `param data` description.

fkokosinski commented on boards/arm/efr32bg_sltb010a/doc/index.rst:
```diff
@@ -163,6 +165,25 @@ the following message:
+
+.. code-block:: console
+
+   west blobs fetch hal_silabs
```
Fixed. Thanks for spotting it.

rettichschnidi commented:
This is probably too much off-topic, but since I have a hard time the get my message beyond the SiLabs FEAs, I try asking here:

As an software engineer I am opposed to any binary blobs running in my applications address space for various reasons (maintainability, debug ability, reliability, etc.). My understanding of the SiLabs blobs/SDK is that they consist of two parts: A FW blob for the 2nd MCU and an abstraction layer layer which runs in my address space (uploading the firmware, communicate with the 2nd MCU). Any chance that SiLabs would release the sources for the abstraction layer, removing the need to run any blobs in my address space?

From a developer (my) perspective, this would make a decision for SiLabs hardware much more likely.

fkokosinski commented:
Hi @rettichschnidi, thanks for the review! Could you PTAL at this PR? The one failing job is not caused by this PR and is being addressed in a separate PR.

rettichschnidi commented on drivers/bluetooth/hci/slz_hci.c:
```diff
@@ -0,0 +1,150 @@
+		bt_recv(buf);
+		break;
+	default:
+		break;
```
How about an error message in case of unexpected values?

rettichschnidi commented on drivers/bluetooth/hci/slz_hci.c:
```diff
@@ -0,0 +1,150 @@
+
+	sl_rail_util_pa_init();
+
+	sl_btctrl_init_mem(SL_BT_CONTROLLER_BUFFER_MEMORY);
```
What if this returns zero? According to the public documentation, this seems possible.

rettichschnidi commented on drivers/bluetooth/hci/slz_hci.c:
```diff
@@ -0,0 +1,150 @@
+	sl_btctrl_hci_parser_init_adv();
+	sl_btctrl_hci_parser_init_conn();
+
+	sl_btctrl_init_basic(SL_BT_CONFIG_MAX_CONNECTIONS,
```
This function returns sl_status_t. Is this guaranteed to succeed? If so, please add a note, otherwise handle the potential errors.

fkokosinski replied:
Added retval checking.

rettichschnidi commented on drivers/bluetooth/hci/slz_hci.c:
```diff
@@ -0,0 +1,150 @@
+
+	sl_rail_util_pa_init();
+
+	sl_btctrl_init_mem(SL_BT_CONTROLLER_BUFFER_MEMORY);
```
What if this returns zero? According to the public documentation, this seems possible.

I think errors should print a log message and return with something else than 0.

fkokosinski replied:
Added retval checking.

rettichschnidi commented on drivers/bluetooth/hci/slz_hci.c:
```diff
@@ -0,0 +1,150 @@
+		bt_recv(buf);
+		break;
+	default:
+		break;
```
It seems to me like unexpected values in data[0] will get swallowed silently, returning zero (success).

How about an error message for such values, and returning a nonzero value?

fkokosinski replied:
Added an error message printing (through the log subsystem) here.

fkokosinski commented on drivers/bluetooth/hci/slz_hci.c:
```diff
@@ -0,0 +1,150 @@
+	sl_rail_util_pa_init();
+
+	sl_btctrl_init_mem(SL_BT_CONTROLLER_BUFFER_MEMORY);
+	sl_btctrl_init_ll();
```
Added retval checking.

fkokosinski commented on drivers/bluetooth/hci/slz_hci.c:
```diff
@@ -0,0 +1,181 @@
+{
+	ARG_UNUSED(unused);
+
+	bt_hci_driver_register(&drv);
```
Done

fkokosinski commented on drivers/bluetooth/hci/slz_hci.c:
```diff
@@ -0,0 +1,181 @@
+	ret = sl_btctrl_init_mem(SL_BT_CONTROLLER_BUFFER_MEMORY);
+	if (!ret) {
+		LOG_ERR("Failed to allocate memory %d", ret);
+		return ret;
```
Good idea. Done.

fkokosinski commented on drivers/bluetooth/hci/slz_hci.c:
```diff
@@ -0,0 +1,181 @@
+	rail_isr_installer();
+	sl_rail_util_pa_init();
+
+	/* sl_btctrl_init_mem returns the number of memory buffers allocated */
+	ret = sl_btctrl_init_mem(SL_BT_CONTROLLER_BUFFER_MEMORY);
+	if (!ret) {
```
> What is the relationship between SL_BT_CONTROLLER_BUFFER_MEMORY (Unit: bytes?) and the returned number (Unit: memory buffers?)?

For BG22, the relationship between the number of bytes the user wants to allocate and the returned memory buffers that were allocated seems to follow `mem_bufs = floor(num_bytes/84)`. 

> Is every return value other than zero actually fine? Even a very small number like 1?

No - it depends on the BT role that is selected in Zephyr. `bluetooth/central` requires at least 54 buffers, whereas `bluetooth/observer` requires at least 27 buffers. The default value set in Kconfig provides a safe margin.

fkokosinski committed:
drivers/bluetooth/hci: add SiLabs BLE HCI driver

This commit adds the SiLabs Bluetooth HCI driver. It also enables this
BLE HCI driver on the efr32bg_sltb010a board.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski commented:
Hi @jhedberg, @jori-nordic! Could you PTAL at this PR? Thanks!

jori-nordic reviewed (dismissed):
I don't have access to a board to test, but looks ok

fkokosinski commented on drivers/bluetooth/hci/slz_hci.c:
```diff
@@ -0,0 +1,185 @@
+ * @brief Transmit HCI message using the currently used transport layer.
+ * The HCI calls this function to transmit a full HCI message.
+ * @param[in] data Packet type followed by HCI packet data.
+ * @param[in] len Length of the data.
```
Fixed.

rettichschnidi reviewed (approved):
LGTM (given there are binary blobs), but I do not have access to the relevant hardware.

fkokosinski commented:
Hi @jhedberg, sorry for pinging again. I had to rebase and force push here because the changes to the HAL were merged. Could you re-ack this PR? Thanks!
